mcu/src/dac.c

```diff
@@ -0,0 +1,75 @@
+#include <msp430.h>
+
+#include "dac.h"
+
+#include <stdio.h>
+
+void init_dac(struct dac_t dac)
+{
+    // I used section 2.5 of the following app note to figure out
+    // how to configure the "smart analog combo" to operate in DAC mode
+    // https://www.ti.com/lit/an/slaa833a/slaa833a.pdf
+
+    // Define the register base addresses for the particular SACx module
+    // that the dac struct is set up for. The base_addr in the dac struct
+    // defines the base address of the SAC module. The register offsets for
+    // each module are the same, so we use the SAC0 offset defintions, which
+    // are defined in msp420fr2355.h
+    uint16_t *DAC_REG = (uint16_t *) (dac.sac_base_addr + OFS_SAC0DAC);
+    uint16_t *OA_REG = (uint16_t *) (dac.sac_base_addr + OFS_SAC0OA);
+    uint16_t *PGA_REG = (uint16_t *) (dac.sac_base_addr + OFS_SAC0PGA);
+
+    // Define the base addresses for the port multiplexing registers, PxSEL1 and
+    // PxSEL0. These offsets are the same for all ports, and are defined in
+    // msp430fr2355.h. We use the port 1 offsets.
+    uint16_t *PX_SEL0 = (uint16_t *) (dac.port_base_addr + OFS_P1SEL0); 
+    uint16_t *PX_SEL1 = (uint16_t *) (dac.port_base_addr + OFS_P1SEL1); 
+
+    // Enable the 1.5 V internal reference
+    /* NOTE: the MSP430FR2355 internal reference defaults to 1.5 V, but
+             but can be changed to 2.0 or 2.5V. */
+    // This will be enabled each time a DAC is initialized, but that's okay
+    // because the bit will always be a 1 after the first initialization.
+    PMMCTL2 |= INTREFEN;
+
+    // Set the DAC reference to the internal 1.5 V reference
+    *DAC_REG |= DACSREF_1;
+
+    // Multiplex the pin to select the DAC output
+    // The PxSELx is set to 0b11 to enable the DAC output on that pin.
+    // The bit # in PxSEL corresponds to the pin # on that port. 
+    // Both SEL1[pin #] and SEL0[pin #] need to be set to a 1.
+    *PX_SEL1 |= dac.port_bit;
+    *PX_SEL0 |= dac.port_bit;
+
+    // Enable the DACs
+    *DAC_REG |= DACEN;
+
+    // Select the DAC as the positive input source
+    *OA_REG |= PSEL_1;
+
+    // Select the PGA source for the negative input; the PGA will be configured for buffer mode
+    *OA_REG |= NSEL_1;
+
+    // Set the PGA to buffer mode
+    *PGA_REG |= MSEL_1;
+
+    // Enable the positive input and negative input muxes
+    *OA_REG |= PMUXEN | NMUXEN;
+
+    // Enable the op-amp outputs
+    *OA_REG |= OAEN;
+
+    // Enable smart analog combo (SAC) module
+    *OA_REG |= SACEN;
+}
+
+// set_dac_voltage() will convert the voltage into the correct 12-bit number, 
+// then write that to the register.
+void set_dac_voltage(struct dac_t dac, uint16_t data)
+{
+    uint16_t scaled_data = (float)data * 2.73;
```

Comment:
We should use the full precision of the conversion constant instead of rounding. We are using the 1.5 V internal reference.

Reply:
are you worried about precision to the degree of using a double precision floating point? if not just a float should be enough

Reply:
Good question. Our voltage range is from 0--1500 mV, and the dac data is 12-bits (the top-4 MSB are unused in the 16-bit register). 

The DAC voltage resolution is 0.3662 mV/bit, but we limit our input resolution to 1 mV (the parser only accepts 4-digit voltage values in mV, e.g. 0123). So we already lose precision because of that (e.g., the closet voltage to 1 mV we can achieve is 1.0986 mV, though of course the DAC itself isn't *that* accurate).

However, it turns out that 2.73 is actually the exact conversion value, which can be exactly represented in either single- or double-precision. My calculation was wrong. I was doing (2^12 / 1500), but it's actually (2^12 - 1)/1500 because the largest 12-bit unsigned integer is 4095, not 4096. So all is well. The conversion constant doesn't need to be updated, but we should change it to a compile-time-calculated constant instead of a hard-coded magic number (for code readability and documentation purposes)

Reply:
Resolved in commit 98adf31518cdac69b4037d63e46052e628735e94